godot-ffi/src/global.rs

```diff
@@ -131,57 +103,52 @@ impl<T> Global<T> {
 
 // Encapsulate private fields.
 mod global_guard {
+    use super::*;
     use std::ops::{Deref, DerefMut};
-    use std::sync::MutexGuard;
-
-    use super::InitState;
 
     /// Guard that temporarily gives access to a `Global<T>`'s inner value.
     pub struct GlobalGuard<'a, T> {
-        // Safety invariant: Is `Initialized`.
-        mutex_guard: MutexGuard<'a, InitState<T>>,
+        // Safety invariant: `OnceCell` has been init.
+        mutex_guard: MutexGuard<'a, OnceCell<T>>,
     }
 
     impl<'a, T> GlobalGuard<'a, T> {
-        pub(super) fn new(mutex_guard: MutexGuard<'a, InitState<T>>) -> Option<Self> {
-            match &*mutex_guard {
-                InitState::Initialized(_) => Some(Self { mutex_guard }),
+        pub(super) fn new(mutex_guard: MutexGuard<'a, OnceCell<T>>) -> Option<Self> {
+            // Use an eager map instead of `mutex_guard.get().map(|_| Self { mutex_guard })`
+            // as `.get().map(…)` trys to move `mutex_guard` while borrowing an ignored value.
+            match mutex_guard.get() {
+                Some(_) => Some(Self { mutex_guard }),
                 _ => None,
             }
         }
 
         /// # Safety
         ///
-        /// The value must be `Initialized`.
-        pub(super) unsafe fn new_unchecked(mutex_guard: MutexGuard<'a, InitState<T>>) -> Self {
-            debug_assert!(matches!(*mutex_guard, InitState::Initialized(_)));
-
+        /// The value must be initialized.
+        pub(super) unsafe fn new_unchecked(mutex_guard: MutexGuard<'a, OnceCell<T>>) -> Self {
+            debug_assert!(mutex_guard.get().is_some(), "safety precondition violated");
             Self::new(mutex_guard).unwrap_unchecked()
         }
     }
 
     impl<T> Deref for GlobalGuard<'_, T> {
         type Target = T;
-
         fn deref(&self) -> &Self::Target {
-            // SAFETY: `self` is `Initialized`.
-            unsafe { self.mutex_guard.as_initialized().unwrap_unchecked() }
+            // SAFETY: [`GlobalGuard::new`] being the sole constructor guarantees that the cell is initialized.
```

Comment:
This isn't quite accurate; there is another constructor `new_unchecked()` with the same visibility.

But I think it's enough to mention that the invariant of `GlobalGuard` is that the cell is initialized. Then ensuring this (and proving its truth via comments) can be done in the constructors.